sdk/core/azure-core/src/main/java/com/azure/core/implementation/ByteCountingAsynchronousByteChannel.java

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import com.azure.core.util.ProgressReporter;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousByteChannel;
+import java.nio.channels.CompletionHandler;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Count bytes written and read to the target channel.
+ */
+public class ByteCountingAsynchronousByteChannel implements AsynchronousByteChannel {
+
+    private final AsynchronousByteChannel channel;
+    private final AtomicLong bytesWritten = new AtomicLong();
+    private final AtomicLong bytesRead = new AtomicLong();
```

Comment:
How commonly do we think these Channel types will be getting used? Very commonly or semi-frequently? `Atomic*`s are known to have much higher overhead than their CAS equivalents.

Reply:
I'll do this in final PRs against main.

Reply:
This has been addressed.